lib/os/timeutil.c

```diff
@@ -56,6 +56,6 @@ time_t timeutil_timegm(struct tm *tm)
 	s64_t ndays = time_days_from_civil(y, m, d);
 
 	return (time_t)tm->tm_sec
-	       + 60 * (tm->tm_min + 60 * tm->tm_hour)
+	       + (time_t)60 * ((time_t)tm->tm_min + 60 * (time_t)tm->tm_hour)
 	       + 86400 * ndays;
```

Comment:
The changed code still violates the rule.

Reply:
can you elaborate it ?
the casts in `((time_t)tm->tm_min + 60 * (time_t)tm->tm_hour)` ensure that it will result in a `time_t` operand,

Reply:
As long as you don't care about dates outside the range `[1970-01-01T00:00:00Z, 2038-01-19T03:14:07Z]` `time_t` can be a signed 32-bit integer.

Reply:
if that is the case, some app bring its own definition of time_t (since zephry's define it to int64) than `ndays` should be a time_t as well.

Reply:
Zephyr's minimal libc *currently* uses `int64_t` for `time_t`.  That may in the future become controllable at build-time, to reduce memory usage.

Other toolchains may also use `int32_t`.  This isn't something the application can control, and it's not something a naive MISRA check can validate.  It's toolchain-dependent.

No, `ndays` shouldn't be `time_t`.  It's not a count of seconds relative to the POSIX epoch.

However, this discussion does convince me that we need to change this API to remove the implementation-defined behavior.  #18465.

Reply:
if that is the case the function is broken as you mentioned, because even making ndays int64 if time_t is a 32 bits variable this function will be wrong.